woocommerce/woocommerce-android
# Card-Present payments: Establishing connection to a mocked reader

Parent issue #3725 

Changes in this PR
1. Initializes CardReaderManager in WooCommerceDebug application class. The feature is in active development and won't be released in the near future. The initialisation is using a static `getInstance` method which lives in `CardReaderManager`. This solution isn't optimal, but I believe it's good enough for this stage of development.
2. Adds logic into `CardReaderSettingsFragment`. This class won't be used in the current state in production. Most of the code will be moved to a viewModel and it'll be fully tested. However, due to the nature of this project, we want to implement all the critical parts of the flow as soon as possible to discover any blockers sooner in the cycle. We'll keep iterating on this class and we'll improve it into a releasable state in upcoming PRs. In other words, don't worry to much about review of this fragment.
3. Updates `CardReaderManager/CardReaderManagerImpl` 
           - Adds `startDiscover` and `connectToReader` methods to the public interface.
           - Adds `discoveryEvents` and `readerStatus` StateFlows to the public interface.
           - Implements the above in `CardReaderManagerImpl`. 
           
The simplified flow is
1. Client (WCAndroid app) subscribes to `discoveryEvents`
2. Client invokes `startDiscovery`
3. `CardReaderManager` updates `discoveryEvents` with all found readers
4. Client receives `ReadersFound` event
5. Client subscribes to `readerStatus`
6. Client invokes `connectToReader` and provides ID which received in `ReaderFound` event in step 4. In the future, we'll display a list of all found readers on the UI and the user will pick the reader they want to connect to. For testing purposes, the app always connects to the first found reader now.
7. `CardReaderManager` updates `readerStatus` with the corresponding connection status.
8. Client receives `CONNECTING/CONNECTED/NOT_CONNECTED` status and acts accordingly - for testing purposes, the app updates a dummy text view with the connection status.


To test:
Testing this PR is a bit tricky, since it requires a running StripeTerminal backend. I'm not sure if it's worth the effort since this feature is not production ready anyway and we'll test it many times before it's released. As soon as we implement the corresponding methods in FluxC the running StripeTerminal backend won't be required anymore. I'll provide the testing instructions, but feel free to skip the tests.
1. Follow `Stripe account` and `Local Terminal Backend` instructions mentioned in p91TBi-4zD-p2. If you don't have access to that post, follow the instructions mentioned [here](https://github.com/stripe/example-terminal-backend#running-locally-on-your-machine). They are almost identical, they just lack a few extra tips.
2. Open `CardReaderStoreImpl.kt` and update `backendUrl` with your IP address
3. Run the app
4. Open App Settings (gear icon in the top right corner)
5. Tap on "Manage Card Readers" 
6. Make sure the local backend is running
7. Tap on  "Connect card reader"
8. Notice "Started" and "Reader found" snackbar messages and notice the connection status textView gets updated with the new state
9. The app should end up in "Connected" state and keep it (wait for at least 20s to make sure the connection is stable -> the StripeTerminalSDK does some verification in the background and if it doesn't succeeds it disconnects the reader after +- 10s). In case the connection keeps dropping, check LogCat - if you see `CONNECTION_TOKEN_PROVIDER_ERROR: java.lang.RuntimeException: Creating connection token failed` your IP address is not correct or the server is not running.
10. Kill the server and notice the app updates the connection status to "NOT_CONNECTED"



Update release notes:
